File: rayon/omp/sieve.c
#include <stdio.h>
#include <omp.h>
#include <math.h>
#include <stdlib.h>

#define MARKED 1
#define UNMARKED 0

int primes_serial(int n);
int primes_parallel(int n);

int main(int argc, char** argv)
{
    for(int ii = 2; ii < 8; ii++)
    {
        int n = pow(10, ii);

        double t1 = omp_get_wtime();
        int primes_ser = primes_serial(n);
        double t2 = omp_get_wtime();
        int primes_par = primes_parallel(n);
        double t3 = omp_get_wtime();

        printf("%d %d %d\n", primes_ser, primes_par, n);
        printf("Sequential: %f, Parallel: %f\n", t2 - t1, t3 - t2);
    }
    return 0;
}

int primes_serial(int n)
{
    char * slots = (char *) malloc(n * sizeof(char));

    for(int i = 0; i < n; i++)
    {
        slots[i] = UNMARKED;
    }

    int end = sqrt(n);

    double start_time = omp_get_wtime();
    for(int i = 2; i <= end; i++)
    {
        if(slots[i-1] == MARKED)
        {
            continue;
        }
        else
        {
            #pragma omp parallel for schedule(runtime)
            for(int j = i*i; j <= n; j += i)
            {
                // printf("%d %d\n", j, i);
                slots[j-1] = MARKED;
            }
        }
    }
    
    double end_time = omp_get_wtime();

    int num_primes = 0;
    for(int i = 0; i < n; i++)
    {
        if(slots[i] == UNMARKED)
        {
            num_primes++;
            // printf("%d %s\n", i + 1, "is prime");
        }
    }
    free(slots);
    // printf("%d\n", num_primes);
    return num_primes;
}

int primes_parallel(int n)
{
    char * slots = (char *) malloc(n * sizeof(char));
    for(int i = 0; i < n; i++)
    {
        slots[i] = UNMARKED;
    }

    int end = sqrt(n);

    for(int i = 2; i <= end; i++)
    {
        if(slots[i-1] == MARKED)
        {
            continue;
        }
        else
        {

            #pragma omp parallel for schedule(runtime)
            for(int j = i*i; j <= n; j += i)
            {
                slots[j-1] = MARKED;
            }
        }
    }

    int num_primes = 0;
    for(int i = 0; i < n; i++)
    {
        if(slots[i] == UNMARKED)
            num_primes++;
    }
    free(slots);
    return num_primes;
}

File: rayon/src/sieve.rs
use rayon::prelude::*;
use std::time::{SystemTime, UNIX_EPOCH};

pub fn sieve()
{
    for ii in 2..8
    {
        let n = 10usize.pow(ii);

        let t1 = since_the_epoch();
        let primes_ser = primes_serial(n);
        let t2 = since_the_epoch();
        let primes_par = primes_parallel(n);
        let t3 = since_the_epoch();

        println!("{} {} {}", primes_ser, primes_par, n);
        println!("Sequential: {}, Parallel: {}", t2 - t1, t3 - t2);
    }
}

const unmarked: bool = false;
const marked: bool = true;

fn primes_serial(n: usize) -> usize
{
    let mut slots: Vec<_> = (0..n).map(|x| unmarked).collect();
    let end = (n as f64).sqrt() as usize;

    for i in 2..(end+1)
    {
        if slots[i-1] == marked
        {
            continue;
        }
        else
        {
            for j in ((i*i)..n).step_by(i)
            {
                slots[j-1] = marked;
            }
        }
    }

    let num_primes = slots.into_iter().filter(|x| *x == unmarked).count();
    num_primes
}

fn primes_parallel(n: usize) -> usize
{
    let mut slots: Vec<_> = (0..n).map(|x| unmarked).collect();
    let end = (n as f64).sqrt().ceil() as usize;

    for i in 2..(end+1)
    {
        if slots[i-1] == marked
        {
            continue;
        }
        else
        {
            // for j in ((i*i)..n).step_by(i)
            (((i*i)..n).step_by(i))
            .collect::<Vec<usize>>()
            .into_par_iter()
            .for_each(|j|
            {
                slots[j-1] = marked;
            });
        }
    }

    let num_primes = slots.into_iter().filter(|x| *x == unmarked).count();
    num_primes
}

fn since_the_epoch() -> u64 {
    let start = SystemTime::now();
    let since_the_epoch = start.duration_since(UNIX_EPOCH).expect(
        "Time went backwards",
    );
    since_the_epoch.as_secs() * 1000 + since_the_epoch.subsec_nanos() as u64 / 1_000_000
}
